cpp/src/arrow/dataset/dataset_test.cc

```diff
@@ -58,5 +59,33 @@ TEST_F(TestSimpleDataSource, GetFragments) {
   AssertDataSourceEquals(reader.get(), &source);
 }
 
+class TestDataset : public DatasetFixtureMixin {};
+
+TEST_F(TestDataset, TrivialScan) {
+  constexpr int64_t kNumberFragments = 4;
+  constexpr int64_t kNumberBatches = 16;
+  constexpr int64_t kBatchSize = 1024;
+
+  auto s = schema({field("i32", int32()), field("f64", float64())});
+  auto batch = ConstantArrayGenerator::Zeroes(kBatchSize, s);
+
+  std::vector<std::shared_ptr<RecordBatch>> batches{kNumberBatches, batch};
+  auto fragment = std::make_shared<SimpleDataFragment>(batches);
+  DataFragmentVector fragments{kNumberFragments, fragment};
+
+  std::vector<std::shared_ptr<DataSource>> sources = {
+      std::make_shared<SimpleDataSource>(fragments),
+      std::make_shared<SimpleDataSource>(fragments),
+  };
+
+  const int64_t total_batches = sources.size() * kNumberBatches * kNumberBatches;
```

Comment:
Hmm... shouldn't that be `sources.size() * kNumberBatches * kNumberFragments`? We have 2 sources, each with kNumberFragments, each with kNumberBatches.